File: mobile-app-uniapp/src/services/p2p/message-manager.js
/**
 * 消息管理器 - 移动端版本
 *
 * 功能：
 * - 消息ID生成和管理
 * - 消息去重（基于ID）
 * - 批量消息发送
 * - 消息确认和重传
 */

class MessageManager {
  constructor(options = {}) {
    this.options = {
      batchSize: options.batchSize || 10,
      batchInterval: options.batchInterval || 100,
      deduplicationWindow: options.deduplicationWindow || 60000,
      maxDeduplicationCache: options.maxDeduplicationCache || 10000,
      enableRetry: options.enableRetry !== false,
      maxRetries: options.maxRetries || 3,
      retryInterval: options.retryInterval || 1000,
      ...options
    }

    // 消息队列
    this.outgoingQueue = new Map()
    this.batchTimers = new Map()

    // 去重缓存
    this.receivedMessages = new Map()
    this.sentMessages = new Map()

    // 待确认消息
    this.pendingAcks = new Map()

    // 事件监听器
    this.eventListeners = new Map()

    // 统计信息
    this.stats = {
      messagesSent: 0,
      messagesReceived: 0,
      messagesDuplicated: 0,
      messagesRetried: 0,
      batchesSent: 0
    }

    // 定期清理
    this.startCleanupTimer()
  }

  /**
   * 生成消息ID
   */
  generateMessageId() {
    const timestamp = Date.now()
    const random = Math.random().toString(36).substring(2, 10)
    const counter = this.stats.messagesSent

    return `${timestamp}-${random}-${counter}`
  }

  /**
   * 发送消息
   */
  async sendMessage(peerId, payload, options = {}) {
    const {
      priority = 'normal',
      requireAck = false,
      immediate = false
    } = options

    const messageId = this.generateMessageId()

    const message = {
      id: messageId,
      type: payload.type,
      payload: payload,
      timestamp: Date.now(),
      priority,
      requireAck
    }

    this.sentMessages.set(messageId, {
      message,
      peerId,
      sentAt: Date.now(),
      retries: 0
    })

    if (priority === 'high' || immediate) {
      await this.sendImmediately(peerId, message)
    } else {
      this.queueMessage(peerId, message)
    }

    if (requireAck) {
      this.waitForAck(messageId, peerId, message)
    }

    this.stats.messagesSent++

    return messageId
  }

  /**
   * 接收消息
   */
  async receiveMessage(peerId, message) {
    const { id, type, payload, requireAck } = message

    if (this.isDuplicate(id)) {
      console.log('[MessageManager] 重复消息，已忽略:', id)
      this.stats.messagesDuplicated++
      return null
    }

    this.receivedMessages.set(id, Date.now())

    if (requireAck) {
      await this.sendAck(peerId, id)
    }

    this.stats.messagesReceived++

    this.emit('message', {
      peerId,
      messageId: id,
      type,
      payload,
      timestamp: message.timestamp
    })

    return {
      id,
      type,
      payload
    }
  }

  /**
   * 接收批量消息
   */
  async receiveBatchMessages(peerId, messages) {
    console.log(`[MessageManager] 接收批量消息: ${messages.length}条`)

    const results = []
    for (const message of messages) {
      try {
        const result = await this.receiveMessage(peerId, message)
        if (result) {
          results.push(result)
        }
      } catch (error) {
        console.error('[MessageManager] 处理消息失败:', error)
      }
    }

    return results
  }

  /**
   * 检查是否重复
   */
  isDuplicate(messageId) {
    return this.receivedMessages.has(messageId)
  }

  /**
   * 加入批量队列
   */
  queueMessage(peerId, message) {
    if (!this.outgoingQueue.has(peerId)) {
      this.outgoingQueue.set(peerId, [])
    }

    const queue = this.outgoingQueue.get(peerId)
    queue.push(message)

    if (queue.length >= this.options.batchSize) {
      this.flushQueue(peerId)
    } else {
      this.scheduleBatchSend(peerId)
    }
  }

  /**
   * 计划批量发送
   */
  scheduleBatchSend(peerId) {
    if (this.batchTimers.has(peerId)) {
      clearTimeout(this.batchTimers.get(peerId))
    }

    const timer = setTimeout(() => {
      this.flushQueue(peerId)
    }, this.options.batchInterval)

    this.batchTimers.set(peerId, timer)
  }

  /**
   * 刷新队列
   */
  async flushQueue(peerId) {
    const queue = this.outgoingQueue.get(peerId)
    if (!queue || queue.length === 0) return

    console.log(`[MessageManager] 批量发送 ${queue.length} 条消息`)

    if (this.batchTimers.has(peerId)) {
      clearTimeout(this.batchTimers.get(peerId))
      this.batchTimers.delete(peerId)
    }

    try {
      await this.sendBatch(peerId, queue)
      this.stats.batchesSent++
    } catch (error) {
      console.error('[MessageManager] 批量发送失败:', error)
    }

    this.outgoingQueue.set(peerId, [])
  }

  /**
   * 立即发送
   */
  async sendImmediately(peerId, message) {
    console.log('[MessageManager] 立即发送消息:', message.id)

    this.emit('send', {
      peerId,
      message
    })
  }

  /**
   * 发送批量消息
   */
  async sendBatch(peerId, messages) {
    console.log(`[MessageManager] 发送批量消息: ${messages.length}条`)

    this.emit('send-batch', {
      peerId,
      messages
    })
  }

  /**
   * 发送确认
   */
  async sendAck(peerId, messageId) {
    const ackMessage = {
      id: this.generateMessageId(),
      type: 'message:ack',
      payload: {
        ackFor: messageId
      },
      timestamp: Date.now()
    }

    await this.sendImmediately(peerId, ackMessage)
  }

  /**
   * 等待确认
   */
  waitForAck(messageId, peerId, message) {
    const timeout = setTimeout(() => {
      console.warn('[MessageManager] 消息确认超时:', messageId)
      this.handleAckTimeout(messageId, peerId, message)
    }, this.options.retryInterval * this.options.maxRetries)

    this.pendingAcks.set(messageId, {
      message,
      peerId,
      timer: timeout
    })
  }

  /**
   * 接收确认
   */
  receiveAck(ackFor) {
    if (this.pendingAcks.has(ackFor)) {
      const { timer } = this.pendingAcks.get(ackFor)
      clearTimeout(timer)
      this.pendingAcks.delete(ackFor)

      console.log('[MessageManager] 收到消息确认:', ackFor)
      this.sentMessages.delete(ackFor)
    }
  }

  /**
   * 处理确认超时
   */
  async handleAckTimeout(messageId, peerId, message) {
    const sentInfo = this.sentMessages.get(messageId)
    if (!sentInfo) return

    if (sentInfo.retries < this.options.maxRetries) {
      console.log(`[MessageManager] 重试发送 (${sentInfo.retries + 1}/${this.options.maxRetries}):`, messageId)

      sentInfo.retries++
      sentInfo.sentAt = Date.now()

      await this.sendImmediately(peerId, message)
      this.stats.messagesRetried++

      this.waitForAck(messageId, peerId, message)
    } else {
      console.error('[MessageManager] 消息发送失败:', messageId)

      this.emit('send-failed', {
        messageId,
        peerId,
        message
      })

      this.sentMessages.delete(messageId)
      this.pendingAcks.delete(messageId)
    }
  }

  /**
   * 启动清理定时器
   */
  startCleanupTimer() {
    setInterval(() => {
      this.cleanupExpiredData()
    }, 60000)
  }

  /**
   * 清理过期数据
   */
  cleanupExpiredData() {
    const now = Date.now()
    let cleanedCount = 0

    for (const [messageId, timestamp] of this.receivedMessages.entries()) {
      if (now - timestamp > this.options.deduplicationWindow) {
        this.receivedMessages.delete(messageId)
        cleanedCount++
      }
    }

    if (this.receivedMessages.size > this.options.maxDeduplicationCache) {
      const toDelete = this.receivedMessages.size - this.options.maxDeduplicationCache
      const entries = Array.from(this.receivedMessages.entries())
        .sort((a, b) => a[1] - b[1])
        .slice(0, toDelete)

      entries.forEach(([messageId]) => {
        this.receivedMessages.delete(messageId)
        cleanedCount++
      })
    }

    if (cleanedCount > 0) {
      console.log(`[MessageManager] 清理 ${cleanedCount} 条过期消息`)
    }
  }

  /**
   * 事件监听
   */
  on(event, callback) {
    if (!this.eventListeners.has(event)) {
      this.eventListeners.set(event, [])
    }
    this.eventListeners.get(event).push(callback)
  }

  /**
   * 触发事件
   */
  emit(event, data) {
    const listeners = this.eventListeners.get(event)
    if (listeners) {
      listeners.forEach(callback => callback(data))
    }
  }

  /**
   * 获取统计信息
   */
  getStats() {
    return {
      ...this.stats,
      queuedMessages: Array.from(this.outgoingQueue.values())
        .reduce((sum, queue) => sum + queue.length, 0),
      pendingAcks: this.pendingAcks.size,
      deduplicationCacheSize: this.receivedMessages.size
    }
  }

  /**
   * 清理资源
   */
  cleanup() {
    for (const timer of this.batchTimers.values()) {
      clearTimeout(timer)
    }
    this.batchTimers.clear()

    for (const { timer } of this.pendingAcks.values()) {
      clearTimeout(timer)
    }
    this.pendingAcks.clear()

    this.outgoingQueue.clear()
    this.receivedMessages.clear()
    this.sentMessages.clear()
  }
}

// 导出单例
let messageManagerInstance = null

export function getMessageManager(options) {
  if (!messageManagerInstance) {
    messageManagerInstance = new MessageManager(options)
  }
  return messageManagerInstance
}

export default MessageManager

File: ios-app/ChainlessChain/Features/Social/Services/MessageManager.swift
import Foundation
import CoreCommon

/// Message Manager - Handles message deduplication, batching, and delivery
/// Reference: desktop-app-vue/src/main/p2p/message-manager.js
@MainActor
class MessageManager: ObservableObject {
    static let shared = MessageManager()

    private let logger = Logger.shared

    // Configuration
    private let batchSize: Int = AppConfig.P2P.batchSize
    private let batchInterval: TimeInterval = AppConfig.P2P.batchInterval
    private let deduplicationWindow: TimeInterval = AppConfig.P2P.deduplicationWindow
    private let maxDeduplicationCache: Int = AppConfig.P2P.maxDeduplicationCache

    // Message queues
    private var outgoingQueues: [String: [Message]] = [:] // peerId -> messages
    private var batchTimers: [String: Timer] = [:]

    // Cleanup timer
    private var cleanupTimer: Timer?

    // Deduplication cache
    private var receivedMessages: [String: Date] = [:] // messageId -> timestamp
    private var sentMessages: [String: SentMessage] = [:] // messageId -> message info

    // Statistics
    @Published var stats = MessageStats()

    struct MessageStats {
        var messagesSent: Int = 0
        var messagesReceived: Int = 0
        var messagesDuplicated: Int = 0
        var batchesSent: Int = 0
    }

    struct Message: Codable, Identifiable {
        let id: String
        let type: MessageType
        let payload: String
        let timestamp: Date
        let priority: Priority
        let requireAck: Bool

        enum MessageType: String, Codable {
            case text
            case image
            case audio
            case video
            case file
            case system
        }

        enum Priority: String, Codable {
            case high
            case normal
            case low
        }
    }

    struct SentMessage {
        let message: Message
        let peerId: String
        let sentAt: Date
        var retries: Int
    }

    private init() {
        // Start cleanup timer
        startCleanupTimer()
    }

    // MARK: - Message Sending

    /// Send message with deduplication and batching
    func sendMessage(
        to peerId: String,
        payload: String,
        type: Message.MessageType = .text,
        priority: Message.Priority = .normal,
        requireAck: Bool = false,
        immediate: Bool = false
    ) async -> String {
        // Generate message ID
        let messageId = generateMessageId()

        // Build message
        let message = Message(
            id: messageId,
            type: type,
            payload: payload,
            timestamp: Date(),
            priority: priority,
            requireAck: requireAck
        )

        // Record sent message
        sentMessages[messageId] = SentMessage(
            message: message,
            peerId: peerId,
            sentAt: Date(),
            retries: 0
        )

        // High priority or immediate send
        if priority == .high || immediate {
            await sendImmediately(peerId: peerId, message: message)
        } else {
            // Add to batch queue
            queueMessage(peerId: peerId, message: message)
        }

        stats.messagesSent += 1

        return messageId
    }

    /// Receive message with deduplication
    func receiveMessage(from peerId: String, message: Message) -> Bool {
        // Check for duplicates
        if isDuplicate(messageId: message.id) {
            logger.debug("[MessageManager] Duplicate message ignored: \(message.id)")
            stats.messagesDuplicated += 1
            return false
        }

        // Record received message
        receivedMessages[message.id] = Date()

        stats.messagesReceived += 1

        logger.debug("[MessageManager] Message received from \(peerId): \(message.id)")

        return true
    }

    // MARK: - Batching

    /// Queue message for batch sending
    private func queueMessage(peerId: String, message: Message) {
        if outgoingQueues[peerId] == nil {
            outgoingQueues[peerId] = []
        }

        outgoingQueues[peerId]?.append(message)

        // Check if batch size reached
        if let queue = outgoingQueues[peerId], queue.count >= batchSize {
            Task {
                await flushQueue(peerId: peerId)
            }
        } else {
            // Schedule batch send
            scheduleBatchSend(peerId: peerId)
        }
    }

    /// Schedule batch send
    private func scheduleBatchSend(peerId: String) {
        // Cancel existing timer
        batchTimers[peerId]?.invalidate()

        // Create new timer
        let timer = Timer.scheduledTimer(withTimeInterval: batchInterval, repeats: false) { [weak self] _ in
            Task { @MainActor in
                await self?.flushQueue(peerId: peerId)
            }
        }

        batchTimers[peerId] = timer
    }

    /// Flush queue (batch send)
    private func flushQueue(peerId: String) async {
        guard let queue = outgoingQueues[peerId], !queue.isEmpty else {
            return
        }

        logger.debug("[MessageManager] Batch sending \(queue.count) messages to: \(peerId)")

        // Cancel timer
        batchTimers[peerId]?.invalidate()
        batchTimers.removeValue(forKey: peerId)

        // Send batch
        await sendBatch(peerId: peerId, messages: queue)

        // Clear queue
        outgoingQueues[peerId] = []

        stats.batchesSent += 1
    }

    /// Send immediately
    private func sendImmediately(peerId: String, message: Message) async {
        logger.debug("[MessageManager] Sending message immediately: \(message.id)")

        // Notify P2P manager to send
        NotificationCenter.default.post(
            name: .p2pSendMessage,
            object: nil,
            userInfo: ["peerId": peerId, "message": message]
        )
    }

    /// Send batch
    private func sendBatch(peerId: String, messages: [Message]) async {
        logger.debug("[MessageManager] Sending batch: \(messages.count) messages")

        // Notify P2P manager to send batch
        NotificationCenter.default.post(
            name: .p2pSendBatch,
            object: nil,
            userInfo: ["peerId": peerId, "messages": messages]
        )
    }

    // MARK: - Deduplication

    /// Check if message is duplicate
    private func isDuplicate(messageId: String) -> Bool {
        return receivedMessages[messageId] != nil
    }

    /// Generate message ID
    private func generateMessageId() -> String {
        let timestamp = Date().timeIntervalSince1970
        let random = UUID().uuidString.prefix(8)
        let counter = stats.messagesSent

        return "\(Int(timestamp))-\(random)-\(counter)"
    }

    // MARK: - Cleanup

    /// Start cleanup timer
    private func startCleanupTimer() {
        // Invalidate existing timer if any
        cleanupTimer?.invalidate()

        cleanupTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.cleanupExpiredMessages()
            }
        }
    }

    /// Stop cleanup timer
    func stopCleanupTimer() {
        cleanupTimer?.invalidate()
        cleanupTimer = nil
    }

    /// Cleanup expired messages
    private func cleanupExpiredMessages() {
        let now = Date()

        // Clean received messages
        receivedMessages = receivedMessages.filter { _, timestamp in
            now.timeIntervalSince(timestamp) < deduplicationWindow
        }

        // Limit cache size using LRU eviction
        if receivedMessages.count > maxDeduplicationCache {
            // Sort by timestamp (oldest first) using safe optional binding
            let sorted = receivedMessages.sorted { entry1, entry2 in
                entry1.value < entry2.value
            }
            let keysToRemove = sorted.prefix(receivedMessages.count - maxDeduplicationCache).map { $0.key }
            keysToRemove.forEach { receivedMessages.removeValue(forKey: $0) }
        }

        logger.debug("[MessageManager] Cleanup: \(receivedMessages.count) messages in cache")
    }

    /// Clear all queues
    func clearQueues() {
        outgoingQueues.removeAll()
        batchTimers.values.forEach { $0.invalidate() }
        batchTimers.removeAll()

        logger.debug("[MessageManager] All queues cleared")
    }

    /// Get statistics
    func getStats() -> MessageStats {
        return stats
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let p2pSendMessage = Notification.Name("p2pSendMessage")
    static let p2pSendBatch = Notification.Name("p2pSendBatch")
    static let p2pMessageReceived = Notification.Name("p2pMessageReceived")
}

